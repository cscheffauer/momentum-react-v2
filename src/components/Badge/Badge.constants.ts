import { TYPE } from '@momentum-design/components/dist/components/badge/badge.constants.js';

const TYPES = Object.values(TYPE);

const DEFAULTS = {
  TYPE: 'dot',
} as const;

export { DEFAULTS, TYPES };
