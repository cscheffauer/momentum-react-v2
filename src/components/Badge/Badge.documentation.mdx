<Badge />` is an independent component.

It has different types - read more about them here:
https://momentum-design.github.io/momentum-design/storybook-static/index.html?path=/docs/components-badge--docs
