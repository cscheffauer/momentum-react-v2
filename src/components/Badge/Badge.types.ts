import type { BadgeType } from '@momentum-design/components';
import type IconKeys from '@momentum-design/icons/dist/types/types';
import type { CSSProperties } from 'react';

export interface Props {
  /**
   * Type of the badge, mapping from momentum-design
   * see https://momentum-design.github.io/momentum-design/en/components/ for more information
   */
  type?: BadgeType;

  /**
   * Current counter number
   *
   * When type === counter
   */
  counter?: number;

  /**
   * Max counter number, if `counter` > maxCounter, `+` will be appended
   *
   * When type === counter
   */
  maxCounter?: number;

  /**
   * Max counter number, if `counter` > maxCounter, `+` will be appended
   *
   * When type === icon
   */
  iconName?: IconKeys;

  /**
   * Max counter number, if `counter` > maxCounter, `+` will be appended
   *
   * When type === icon
   */
  variant?: 'primary' | 'secondary';

  /**
   * If the badge should be used as a overlay (having a border)
   * Like placed on a button for indicating notifications.
   */
  overlay?: boolean;

  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;
}
