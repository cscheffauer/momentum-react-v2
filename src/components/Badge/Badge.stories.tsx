/* eslint-disable import/no-unresolved */
import React, { FC } from 'react';
import { Story } from '@storybook/react';
import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs';

import Badge, { BadgeProps, BADGE_CONSTANTS as CONSTANTS } from './';
import Documentation from './Badge.documentation.mdx';
import { TYPES } from './Badge.constants';

const DocsPage: FC = () => (
  <>
    <Title />
    <Subtitle />
    <Description />
    <Documentation />
    <Primary />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  title: 'Momentum UI/Badge',
  component: Badge,
  parameters: {
    expanded: true,
    docs: {
      page: DocsPage,
    },
  },
};

const argTypes = {
  type: {
    description: 'Modifies the type of the badge',
    options: [undefined, ...TYPES],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.TYPE,
      },
    },
  },
  counter: {
    description: 'Provides the counter number for this element. \n\n (when `type` === counter)',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  maxCounter: {
    description: 'Provides the maxCounter number for this element. \n\n (when `type` === counter)',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  iconName: {
    description: 'Provides the iconName for this element. \n\n (when `type` === icon)',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  overlay: {
    description: 'Provides the overlay attribute for this element.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
};

const Template: Story<BadgeProps> = (args: BadgeProps) => {
  return <Badge {...args} />;
};

const Example = Template.bind({});
Example.argTypes = { ...argTypes };

export { Example };
