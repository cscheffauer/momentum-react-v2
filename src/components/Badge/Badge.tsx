import React, { FC } from 'react';
import classnames from 'classnames';

import type { Props } from './Badge.types';
import { DEFAULTS } from './Badge.constants';
import { Badge as MdcBadge } from '@momentum-design/components/dist/react';

const Badge: FC<Props> = (props: Props) => {
  const {
    className,
    id,
    style,
    type = DEFAULTS.TYPE,
    counter,
    maxCounter,
    iconName,
    overlay,
  } = props;

  return (
    <MdcBadge
      type={type}
      counter={counter}
      maxCounter={maxCounter}
      iconName={iconName}
      overlay={overlay}
      className={classnames(className)}
      id={id}
      style={style}
    />
  );
};

export default Badge;
