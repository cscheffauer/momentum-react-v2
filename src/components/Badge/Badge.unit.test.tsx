import React from 'react';
import { render, waitFor } from '@testing-library/react';
import Badge from '.';

describe('Badge', () => {
  let container;

  describe('snapshot', () => {
    const setup = async (component: any) => {
      const { container } = render(component);

      // we have to wait for the web component to be rendered
      await waitFor(() => {
        expect(container.querySelector('mdc-badge')).toBeTruthy();
      });

      return container;
    };
    it('should match snapshot without attributes', async () => {
      expect.assertions(2);
      container = await setup(<Badge className="test" />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = dot', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="dot" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = icon', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="icon" iconName="accessibility-regular" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = counter', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="counter" counter={9} maxCounter={20} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = success', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="success" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = warning', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="warning" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with type = error', async () => {
      expect.assertions(2);
      container = await setup(<Badge type="error" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with overlay = true', async () => {
      expect.assertions(2);
      container = await setup(<Badge overlay />);

      expect(container).toMatchSnapshot();
    });
  });
});
