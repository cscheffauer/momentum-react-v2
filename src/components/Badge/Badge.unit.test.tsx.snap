// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge snapshot should match snapshot with overlay = true 1`] = `
<div>
  <mdc-badge
    class=""
    type="dot"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = counter 1`] = `
<div>
  <mdc-badge
    class=""
    max-counter="20"
    type="counter"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = dot 1`] = `
<div>
  <mdc-badge
    class=""
    type="dot"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = error 1`] = `
<div>
  <mdc-badge
    class=""
    type="error"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = icon 1`] = `
<div>
  <mdc-badge
    class=""
    type="icon"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = success 1`] = `
<div>
  <mdc-badge
    class=""
    type="success"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot with type = warning 1`] = `
<div>
  <mdc-badge
    class=""
    type="warning"
    variant="primary"
  />
</div>
`;

exports[`Badge snapshot should match snapshot without attributes 1`] = `
<div>
  <mdc-badge
    class="test"
    type="dot"
    variant="primary"
  />
</div>
`;
